### ISSUE TITLE
Error when unmarshalling with JAXB

### ISSUE THREAD

**Issue by bartlomiej-kfn** [leaningtech/cheerpj-meta#85]
**CheerpJ version: 2.1
JDK version: 1.8.0_161**

Hi, I have two simple classes:

```
@XmlRootElement
public class Task {
    
    private String id;

    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }
}
```
```
public class Main {
    
    public static void main(String[] args) {
        if(args.length == 0) {
            System.out.println("No arg");
            return;
        }
        String xml = args[0];
        Task task = deserialize(xml);
        System.out.println(String.format("Task id: %s", task.getId()));
    }
    
    
    private static Task deserialize(String xml)  {
        try {
            JAXBContext context = JAXBContext.newInstance(Task.class);
            Unmarshaller unmarshaller = context.createUnmarshaller();
            Task task = (Task) unmarshaller.unmarshal(new StringReader(xml));
            return task;
        } catch(Exception ex) {
            throw new RuntimeException(ex.getMessage(), ex);
        }
    }
    
}
```

when i invoke compiled version like this:

```
  cheerpjInit().then(function() {
                cheerpjRunMain("com.mycompany.hello_world.Main", "/app/my.jar");
            });

```

it prints "No args" as expected, but when I call:

```
  cheerpjInit().then(function() {
                cheerpjRunMain("com.mycompany.hello_world.Main", "/app/my.jar", "<task><id>testId</id></task>");
            });

```

it should print "Task id: testId", but i get following error:

```
loader.js:401 Uncaught ReferenceError: _c4pbQpFpdMGdZdqdwD9bXYIZVLt0Ya$ElGKI_iWvGMGaHKKtK5VbXK4WVurfqW$b_tG9b_Bq8gFBqm5vd is not defined
    at N3com9mycompany11hello_world47Task$JaxbAccessorM_getId_setId_java_lang_StringX (eval at _m4Vanpn3cGKuPdWaruT9arbsdrrsqCTNnn (rt.jar.js:298), <anonymous>:4:652)
    at cheerpjClassInit (loader.js:2470)
    at cheerpjEnsureInitialized (loader.js:2182)
    at N4java4lang5Class._m4F05Class13getSuperce5WWE20 [as v25] (rt.jar.js:241)
    at _c4pbMRatbVAGsdSdJcKgZSubclassOfE9 (rt.jar.js:2055)
    at N3sun7reflect17ReflectionFactory._c4pbMNatcVVAGsdSZ0Factory22newConstruebHYAccessorE5 [as v14] (rt.jar.js:2067)
    at _m4FcmvDnNtJ0ueLWKtKToGaefseXstSSMGaHKKtKNDo (rt.jar.js:586)
    at N4java4lang7reflect11Constructor._m4FcmXDnNtJ0ueLWKtuDDdHLyJ0eqJaHNTn [as v66] (rt.jar.js:586)
    at N4java4lang5Class._m4F15Class11newInstanceE7 [as v12] (rt.jar.js:283)
    at _c4pcQtqNj0i04bind2v27runtimeiWKaAS$ZKno6ZecdsgXGMGaHKKt0hGWKtewDnsJ0eqJacWeH3n (rt.jar.com.sun.xml.internal.bind.v2.runtime.reflect.js:245)
```

Both versions works fine when I execute this in console as `java -jar my.jar`. What is the cause of this problem?

Best regards

**Comment by alexp-sssup** [comment]
A function implementation is missing, you can inspect the demangled name by running the following code in the browser console (please note that the first underscore character should be skipped):

```
cheerpjDecompressSymbol("c4pbQpFpdMGdZdqdwD9bXYIZVLt0Ya$ElGKI_iWvGMGaHKKtK5VbXK4WVurfqW$b_tG9b_Bq8gFBqm5vd")
```

```
"ZN3com9mycompany11hello_world47Task$JaxbAccessorM_getId_setId_java_lang_String3getE1"
```

So an internal method from the library is missing. There are 3 likely explanation, in order of probability

1) The method is a Java "native" method, implemented in the JNI which obviously CheerpJ cannot execute out-of-box. There are solutions to this problem that we can discuss later on.
2) There may be a mistake in the build sequence. In the case you are using the ```--pack-jar``` command line option of ```cheerpjfy.py``` make sure that you are generating the jar.js file from the original jar, and not from the packed jar. The packed jar contains all the metadata of classes, but is stripped of all code.
3) There may be a genuine bug in the CheerpJ AOT compiler.

To know for sure what the problem is, we need to see the JAR files. Can you share with us a self-contained reproducible test case, including all dependencies?

**Comment by bartlomiej-kfn** [comment]
This is just simple maven(3.6.3) project which contains those classes and pom.xml file, it doesn't have any dependencies.
After compiling with maven i run `cheerpjfy.py my.jar` on generated artifact and place both my.jar and my.jar.js in web server. I attach full project below.

[hello_world.zip](https://github.com/leaningtech/cheerpj-meta/files/4980456/hello_world.zip)

**Comment by alexp-sssup** [comment]
Thanks for the test case. We can reproduce the issue and at first sight this may actually be an unsupported corner case in reflection support. Unfortunately we are currently not able to commit time to fix issues from non-commercial users, apologies. 

We are certainly willing to get this fixed as part of a license agreement though.

We have also written this issue down (with additional information) in our internal bug tracker, and we will make sure to let you know when the problem is fixed.

**Comment by HoBu23** [comment]
Same issue:

I have updated NetBeans to version 12. My Application was created in Java 8. I use now java 14. All is running but this:

```
    [java.lang.ClassNotFoundException: com.sun.xml.internal.bind.v2.ContextFactory]
    	at javax.xml.bind.JAXB.unmarshal(JAXB.java:171)
```

I raead many posts in Google but i can’t hit a solution.
This error occurs only if i run in NetBeans app. If i compile my app and run it it works without errors.

I habe also downloaded jaxb-api-2.3.1.jar and added to library.
Edit: Alos tryed jaxb-api-2.4.0-b180830.0359.jar. No success.

But realy crasy is, if i run my app in Terminal i get this error:

`    Exception in thread "main" java.lang.NoClassDefFoundError: javax/xml/bind/JAXB`

Only on click on myApp.jar in my current folder it runs without any error:

What is a solution for this behavior?

For now i found a solution but whith this WARNING:

```
    WARNING: An illegal reflective access operation has occurred
    WARNING: Illegal reflective access by com.sun.xml.bind.v2.runtime.reflect.opt.Injector$1 (file:/Applications/NetBeans/Apache%20NetBeans%2012.0.app/Contents/Resources/NetBeans/netbeans/ide/modules/ext/jaxb/jaxb-impl.jar) to method java.lang.ClassLoader.defineClass(java.lang.String,byte[],int,int)
    WARNING: Please consider reporting this to the maintainers of com.sun.xml.bind.v2.runtime.reflect.opt.Injector$1
    WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
    WARNING: All illegal access operations will be denied in a future release
```

this appears if i use this JAXB 2.2.5.jar library. 
Don’t work wit JAXB versions as described above.

My app runs now with 2.2.5 but how long?

Sorry but i have no idea how i can inform the developer. i.g. Sun or Marven or whoever. I true this issue is known there. But i’s bad that it isn’t fixt yet. :(

This ist a part of my code:

```
    package Tools;
    
    import Enums.Property;
    import java.io.File;
    import java.util.LinkedList;
    import java.util.List;
    import javax.swing.JOptionPane;
    import static Tools.Device.Director;
    import javax.xml.bind.JAXB;
    import javax.xml.bind.annotation.XmlElement;
    import javax.xml.bind.annotation.XmlRootElement;
    
    
        @XmlRootElement
        public class Device {
        
            @Override
            public String toString() {
                return DeviceName;
            }
        
            /**
             *
             */
            public static final String Separator = FileHandler.getFileSeparator();
            public static File Director = new File(Property.datarefBaseFile.getValue().substring(0, Property.datarefBaseFile.getValue().lastIndexOf(Separator)) + Separator + "Devices" + Separator);
        
            @XmlElement
            public String DeviceName;
            @XmlElement
            public String DevicePosX;
            @XmlElement
            public String DevicePosY;
            @XmlElement
            public String DeviceWidth;
            @XmlElement
            public String DeviceHeight;
            @XmlElement
            public String DeviceBackgroundImage;
            @XmlElement
            public String DeviceScaleX;
            @XmlElement
            public String DeviceScaleY;
        
            public List<ControlProps> controlsProps;
        
            public static Device load(String file) {
                return JAXB.unmarshal(new File(Director, file), Device.class);
            }
        
            public static File getDirector() {
                return Director;
            }
        
            public void save() {
                Director.mkdir();
                File file = new File(Director, DeviceName + ".xml");
                File fileback = new File(Director, DeviceName + ".bak");
        
                if (fileback.exists()) {
                    fileback.delete();
                    System.out.println(String.format("Device %s old deleted! %s", fileback, Utils.ShowTime()));
                }
                file.renameTo(fileback);
        
                if (fileback.exists()) {
                    System.out.println(String.format("Device %s new created! %s", fileback, Utils.ShowTime()));
                }
        
                if (!file.exists()) {
                    JAXB.marshal(this, file);
                    String msg = String.format("Device %s saved! %s", file, Utils.ShowTime());
                    System.out.println(msg);
                }
        
                if (!file.exists()) {
                    String msg = String.format("Fatal error: %s was deleted an not created new!\nDevice crashed.\nRestore from %s file. %s", file, fileback, Utils.ShowTime());
                    System.out.println(msg);
                    MsgBox.error(msg);
                }
        
                if (!fileback.exists()) {
                    String msg = String.format("Warning! Backupfile: %s was not created yet! %s", fileback, Utils.ShowTime());
                    System.out.println(msg);
                    MsgBox.error(msg);
                }
                //       JOptionPane.showMessageDialog(null, msg);
                //       LogFile.log(Level.INFO, new String[] {file.toString() + " saved"});
            }
        
            public int delete() {
                File fileBak = new File(Director, DeviceName + ".creater.bak");
                File file = new File(Director, DeviceName + ".xml");
                int response = 0;
                String msg = String.format("Delete this device: %s?", DeviceName);
                response = JOptionPane.showConfirmDialog(null, msg, "Confirm",
                        JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE);
                if (response == JOptionPane.YES_OPTION && file.exists()) {
                    file.renameTo(fileBak);
                    file.delete();
                    msg = String.format("File: %s deleted\nBackupfile %s was created!", file, fileBak);
                    JOptionPane.showMessageDialog(null, msg);
                } else {
                    msg = String.format("%s not deleted file: %s", DeviceName, file);
                    JOptionPane.showMessageDialog(null, msg);
                }
        
                return response;
            }
        
            public Device() {
                controlsProps = new LinkedList<>();
            }
        
            public ControlProps find(String name) {
                for (ControlProps prop : controlsProps) {
        
                    if (prop.ControlName.equals(name)) {
                        return prop;
                    }
                }
                throw new IllegalArgumentException("ControlName not found");
            }
        
            public void deleteControl(ControlProps controlProps) {
                String name = controlProps.ControlName;
                if (controlsProps.remove(controlProps)) {
                    save();
                    System.out.println(name + " deleted");
                    //LogFile.log(Level.INFO, new String[]{name + " deleted"});
                }
            }
            
            
        }
```

**Comment by alexp-sssup** [comment]
@HoBu23 This repository is for bugs related to CheerpJ: A Java-to-HTML5 compiler. We cannot help you with general Java issues.

**Comment by HoBu23** [comment]
But this

> @HoBu23 This repository is for bugs related to CheerpJ: A Java-to-HTML5 compiler. We cannot help you with general Java issues.
ok, i have'nt read this, sorry.

**Comment by alexp-sssup** [comment]
This issue is fixed as part of the new CheerpJ 3.0 architecture
